pkg/arvo/app/chat-hook.hoon

```diff
@@ -0,0 +1,364 @@
+  ::
+      %delete
+    ?>  ?=([* ^] path.diff)
+    =/  shp  (~(get by synced) path.diff)
+    ?~  shp
+      [~ this]
+    ?.  =(u.shp src.bol)
+      [~ this]
+    :_  this(synced (~(del by synced) path.diff))
+    :-  (chat-poke diff)
+    [ost.bol %pull [%mailbox path.diff] [src.bol %chat-hook] ~]~
+  ::
+      %message
+    :_  this
+    ?>  ?=([* ^] path.diff)
+    =/  shp  (~(get by synced) path.diff)
+    ?~  shp  ~
+    ?.  =(src.bol u.shp)  ~
+    [(chat-poke diff)]~
+  ::
+  ==
+::
+++  quit
+  |=  wir=wire
+  ^-  (quip move _this)
+  ~&  chat-hook-quit+wir
+  ?:  =(wir /permissions)
+    :_  this
+    [ost.bol %peer /permissions [our.bol %permission-store] /updates]~
+  ?>  ?=([* ^] wir)
+  ?.  (~(has by synced) t.wir)
+    ::  no-op
+    [~ this]
+  ~&  %chat-hook-resubscribe
+  :_  (track-bone wir)
+  [ost.bol %peer wir [(slav %p i.t.wir) %chat-hook] wir]~
+::
+++  reap
+  |=  [wir=wire saw=(unit tang)]
+  ^-  (quip move _this)
+  ?~  saw
+    [~ this]
+  ?>  ?=(^ wir)
+  ~&  %chat-hook-reap
+  [((slog u.saw) ~) this(synced (~(del by synced) t.wir))]
+::
+++  chat-poke
+  |=  act=chat-action
+  ^-  move
+  [ost.bol %poke / [our.bol %chat-store] [%chat-action act]]
+::
+++  permission-poke
+  |=  act=permission-action
+  ^-  move
+  [ost.bol %poke / [our.bol %permission-store] [%permission-action act]]
+::
+++  create-permission
+  |=  [pax=path sec=chat-security]
+  ^-  (list move)
+  =/  read-perm   (weld pax /read)
+  =/  write-perm  (weld pax /write)
+  ?-  sec
+      %channel
+    :~  (permission-poke (sec-to-perm read-perm %black))
+        (permission-poke (sec-to-perm write-perm %black))
+    ==
+  ::
+      %village
+    :~  (permission-poke (sec-to-perm read-perm %white))
+        (permission-poke (sec-to-perm write-perm %white))
+    ==
+  ::
+      %journal
+    :~  (permission-poke (sec-to-perm read-perm %black))
+        (permission-poke (sec-to-perm write-perm %white))
+    ==
+  ::
+      %mailbox
+    :~  (permission-poke (sec-to-perm read-perm %white))
+        (permission-poke (sec-to-perm write-perm %black))
+    ==
+  ::
+  ==
+::
+++  delete-permission
+  |=  pax=path
+  ^-  (list move)
+  =/  read-perm   (weld pax /read)
+  =/  write-perm  (weld pax /write)
+  :~  (permission-poke [%delete read-perm])
+      (permission-poke [%delete write-perm])
+  ==
+::
+++  sec-to-perm
+  |=  [pax=path =kind]
+  ^-  permission-action
+  [%create pax kind *(set ship)]
+::
+++  chat-scry
+  |=  pax=path
+  ^-  (unit mailbox)
+  =.  pax  ;:(weld /=chat-store/(scot %da now.bol)/mailbox pax /noun)
+  .^((unit mailbox) %gx pax)
+::
+++  permitted-scry
+  |=  pax=path
+  ^-  ?
+  .^(? %gx ;:(weld /=permission-store/(scot %da now.bol)/permitted pax /noun))
+::
+++  track-bone
+  |=  wir=wire
+  ^+  this
+  =/  bnd  (~(get by boned) wir)
+  ?^  bnd
+    this(boned (~(put by boned) wir (snoc u.bnd ost.bol)))
+  this(boned (~(put by boned) wir [ost.bol]~))
+::
+++  pull-wire
+  |=  pax=path
+  ^-  (list move)
+  ?>  ?=([* ^] pax)
```

Comment:
This assertion looks unnecessarily specific: `?=(^ pax)` should be enough as you only ever reference `t.pax`.

Reply:
You're right, thanks.

Reply:
The overly-specific pattern-match is still present.

Reply:
It's now been removed.